packages/stargate/src/feemarket.ts

```diff
@@ -0,0 +1,224 @@
+/* eslint-disable @typescript-eslint/no-use-before-define, no-bitwise */
+import { Decimal } from "@cosmjs/math";
+import { BinaryReader, BinaryWriter } from "cosmjs-types/binary";
+
+import { GasPrice } from "./fee";
+import { decodeCosmosSdkDecFromProto } from "./queryclient/utils";
+
+/**
+ * Configuration for dynamic gas price queries from feemarket modules.
+ */
+export interface DynamicGasPriceConfig {
+  /** Denomination to query gas price for (e.g., "uatom") */
+  readonly denom: string;
+  /** Multiplier to apply to the queried gas price (e.g., 1.1 for 10% above base). Defaults to 1.3 */
+  readonly multiplier?: number;
+  /** Minimum gas price to use as fallback if query fails and as a floor for the dynamic price */
+  readonly minGasPrice: GasPrice;
+  /** Maximum gas price to cap at. If not provided, no maximum is enforced */
+  readonly maxGasPrice?: GasPrice;
+}
+
+export function isDynamicGasPriceConfig(
+  config: GasPrice | DynamicGasPriceConfig,
+): config is DynamicGasPriceConfig {
+  return "minGasPrice" in config;
+}
+
+/**
+ * Multiplies a Decimal by a number multiplier.
+ * This is needed because Decimal.multiply() only accepts integers.
+ *
+ * Uses decimal arithmetic to avoid floating point precision issues.
+ */
+export function multiplyDecimalByNumber(
+  value: Decimal,
+  multiplier: number,
+  fractionalDigits: number,
+): Decimal {
+  // Normalize value to target fractional digits
+  const normalizedValue = value.adjustFractionalDigits(fractionalDigits);
+
+  // Convert multiplier to Decimal with same fractional digits
+  const multiplierDecimal = Decimal.fromUserInput(multiplier.toString(), fractionalDigits);
```

Comment:
I tried `multiplier.toFixed(fractionalDigits)` but it leads to some inconvenient effekts like this:

```
> 0.1.toFixed(18)
'0.100000000000000006'
```

i.e. we swap one problem with another. Maybe it is good enough as-is, assuming users don't put in `multiplier` values that lead to scientific notation.